CLN: replace unnecessary `freq_to_period_freqstr` with `PeriodDtype` constructor

natmokval opened the pull request:
xref #52064

replaced `freq_to_period_freqstr()` with `PeriodDtype(freq)._freqstr` in ValueError message in the definition of `_shift_with_freq()`

natmokval committed:
replace freq_to_period_freqstr() with PeriodDtype(freq)._freqstr in _shift_with_freq()

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
replace freq_to_period_freqstr with PeriodDtype()._freqstr except in frequencies.py, period.pyx

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -11274,9 +11274,9 @@ def _shift_with_freq(self, periods: int, axis: int, freq) -> Self:
                 assert orig_freq is not None  # for mypy
                 raise ValueError(
-                    f"Given freq {freq_to_period_freqstr(freq.n, freq.name)} "
+                    f"Given freq {PeriodDtype(freq)._freqstr} "
```
does this change the message? if so, could you show before vs after please?

if not, then is it possible to do this replacement across the whole codebase and get rid of `freq_to_period_freqstr`?

natmokval replied:
> does this change the message? if so, could you show before vs after please?
> 
It doesn't change the message.
> if not, then is it possible to do this replacement across the whole codebase and get rid of `freq_to_period_freqstr`?
>
I replaced `freq_to_period_freqstr` with `PeriodDtype(offset)._freqstr` where it works. Now I am trying to do this replacement in `pandas/tseries/frequencies.py` and in `pandas/_libs/tslibs/period.pyx`, but I get failures. I am not sure we can complitely rid of `freq_to_period_freqstr`

natmokval committed:
fix pre-commit errors

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
resolve conflict

natmokval committed:
resolve conflict

natmokval committed:
remove freq_to_period_freqstr from _maybe_coerce_freq and from tests

natmokval committed:
remove freq_to_period_freqstr from period.pyx

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
correct raise_on_incompatible

natmokval committed:
correct raise_on_incompatible again

natmokval commented:
I replaced `freq_to_period_freqstr` with `PeriodDtype(obj)._freqstr / PeriodDtypeBase(obj._period_dtype_code, obj.n)._freqstr`. Now we don’t need the `function freq_to_period_freqstr`, which is why I removed it.

One comment:
In `pandas/core/arrays/period.py` in the definition of `raise_on_incompatible` I added
```
with warnings.catch_warnings():
            warnings.simplefilter("ignore")
```
because  otherwise the `test_maybe_convert_timedelta` (in `pandas/tests/indexes/period/test_period.py`) fail.

@MarcoGorelli, could you please take a look at this PR?

MarcoGorelli commented on pandas/core/arrays/period.py:
```diff
@@ -986,13 +985,15 @@ def raise_on_incompatible(left, right) -> IncompatibleFrequency:
         other_freq = None
     elif isinstance(right, BaseOffset):
-        other_freq = freq_to_period_freqstr(right.n, right.name)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
```
looks a bit dangerous :) is this for the business day warning? if so, I'd suggest either enforcing that one first, or using `warnings.filterwarnings` and only filtering the warning you mean to

natmokval replied:
yes, this for the business day warning. I agree, it's better to use `warnings.filterwarnings` instead of `warnings.simplefilter`. I made changes in order to filter: `"Future warning "PeriodDtype[B] is deprecated"`

natmokval replied:
we can rid of this `warnings.filterwarnings` after enforcing depreciation of `PeriodDtype[B]`, or we can postpone merging this PR until enforcing deprecation of PeriodDtype[B] is done.

natmokval committed:
replace warnings.simplefilter with warnings.filterwarnings in raise_on_incompatible

MarcoGorelli commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1722,8 +1716,15 @@ cdef class PeriodMixin:
+            if hasattr(other, "_period_dtype_code"):
+                other_freqstr = PeriodDtypeBase(
+                    other._period_dtype_code, other.n
+                )._freqstr
+            else:
+                other_freqstr = other.freqstr
```
why do we need the if-else branch for `other` but not for `self.freq`?

natmokval replied:
I think because `self.freq` always has attribute `"_period_dtype_code",` bur `other` sometimes doesn't.
For example in many tests in `pandas/tests/arithmetic/test_period.py`( e.g. `test_pi_add_sub_timedeltalike_freq_mismatch_monthly`) other is equal `pd.offsets.YearBegin(2)`

MarcoGorelli commented on pandas/tests/plotting/test_datetimelike.py:
```diff
@@ -253,7 +256,7 @@ def test_line_plot_datetime_frame(self, freq):
             index=idx,
             columns=["A", "B", "C"],
         )
-        freq = freq_to_period_freqstr(1, df.index.freq.rule_code)
+        freq = re.split(r"\d", PeriodDtype(df.index.freq)._freqstr)[-1]
```
why does this one require regex splitting, but others don't?

natmokval replied:
Hmm, seems like we  don't need this regex splitting here and can replace `freq = re.split(r"\d", PeriodDtype(df.index.freq)._freqstr)[-1]` with `freq = PeriodDtype(df.index.freq)._freqstr`

I will make this change.

There is another test `test_line_plot_period_mlt_frame` in which we use the same regex splitting. I tried to do the similar replacement (with `freq = PeriodDtype(df.index.freq)._freqstr[1:]`), but the test failed for `"1s"`.

natmokval replied:
I think in `test_line_plot_period_mlt_frame` we can change `freq`  from
```
    freq = re.split(r"\d", PeriodDtype(df.index.freq)._freqstr)[-1]
    freq = df.index.asfreq(freq).freq
```
to 
```
freq = df.index.freq.rule_code
```
and then get rid of regex splitting. I modified the test.

natmokval replied:
if we use the same replacement in `test_line_plot_datetime_frame`, e.g. replace 
```
freq = PeriodDtype(df.index.freq)._freqstr
freq = df.index.to_period(freq).freq
```
with
```
freq = df.index.freq.rule_code
```
the test passes. Should we modify this test as well?

MarcoGorelli reviewed (commented):
nice one - should be close! just got one more question

natmokval committed:
remove regex splitting from test_line_plot_datetime_frame

natmokval committed:
Merge branch 'main' into depr-correct-rename-M-to-ME-II

natmokval committed:
remove regex splitting from test_line_plot_period_mlt_frame

MarcoGorelli reviewed (approved):
Nice!

@jbrockmendel I think you'd suggested this clean up, fancy taking a quick look?

jbrockmendel commented on pandas/core/arrays/period.py:
```diff
@@ -325,7 +324,7 @@ def _from_datetime64(cls, data, freq, tz=None) -> Self:
         """
         if isinstance(freq, BaseOffset):
-            freq = freq_to_period_freqstr(freq.n, freq.name)
+            freq = PeriodDtype(freq)._freqstr
```
we dont need freq.n here?

jbrockmendel replied:
woops, never mind

mroeschke commented:
Thanks @natmokval
